applications/FluidDynamicsApplication/custom_processes/calulate_levelset_consistent_nodal_gradient_process.cpp

```diff
@@ -0,0 +1,148 @@
+    ModelPart& rModelPart,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModelPart
+    ){}
+
+/// Constructor with Kratos model
+CalulateLevelsetConsistentNodalGradientProcess::CalulateLevelsetConsistentNodalGradientProcess(
+    Model& rModel,
+    Parameters Parameters)
+    : CalulateLevelsetConsistentNodalGradientProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString())
+    ){}
+
+void CalulateLevelsetConsistentNodalGradientProcess::Execute(){
+
+    KRATOS_TRY;
+
+    const auto zero_vector = ZeroVector(3);
+    // Set to zero
+    block_for_each(mrModelPart.Nodes(), [&](Node<3>& rNode){
+        rNode.SetValue(NODAL_AREA, 0.0);
+        rNode.SetValue(PRESSURE_GRADIENT, zero_vector);
+    });
+
+    // Current domain size
+    const unsigned int num_dim = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+    const unsigned int num_nodes = num_dim + 1; //For tetrahedra and triangles
+
+    // Auxiliar containers
+    double detJ0 = 0.0;
+    Matrix DN_DX, J0, InvJ0;
+    Vector N(num_nodes), pressures(num_nodes), distances(num_nodes), grad(num_dim);
+
+    // First element iterator
+    const auto it_element_begin = mrModelPart.ElementsBegin();
+
+    const int elements_number = mrModelPart.Elements().size();
+
+    // Iterate over the elements
+    #pragma omp parallel for firstprivate(DN_DX, N, J0, InvJ0, detJ0, pressures, distances, grad)
+    for(int i_elem = 0; i_elem < elements_number; ++i_elem) {
+
+        const auto it_elem = it_element_begin + i_elem;
+        auto& r_geometry = it_elem->GetGeometry();
+
+        // Current geometry information
+        const unsigned number_of_nodes = r_geometry.PointsNumber();
+
+        for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node){
+            distances(i_node) = r_geometry[i_node].FastGetSolutionStepValue(DISTANCE);
+        }
+
+        unsigned int nneg=0, npos=0;
+        for(unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            if(distances(i) > 0) {
+                npos += 1;
+            } else {
+                nneg += 1;
+            }
+        }
+
+        if(nneg == 0 || npos == 0)
+        {
+            // The integration points
+            const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod();
+            const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method);
+            const unsigned int number_of_integration_points = r_integration_points.size();
+
+            for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node)
+                pressures[i_node] = r_geometry[i_node].FastGetSolutionStepValue(PRESSURE);
+
+            // The containers of the shape functions and the local gradients
+            const auto& rNcontainer = r_geometry.ShapeFunctionsValues(r_integration_method);
+            const auto& rDN_DeContainer = r_geometry.ShapeFunctionsLocalGradients(r_integration_method);
```

Comment:
Why using the geometry and then the `GeometryUtils`? Considering that we assume that this will be exclusively used with triangles and tetras, I'd always use the `GeometryUtils` as these are optimized for such geometries.

Reply:
Could you please provide a little more details? I assume that you suggest using "GeometryUtils::ShapeFunctionsGradients(rDN_De, mInvJ0, rDN_DX0)"?

Reply:
Yes. The point is that you are calculating the gradients and shape functions in these two lines. Then you retrieve the gradients again from the `GeometryUtils`.

Reply:
I thoroughly checked the functions; the current implementation seems to have no duplicate calls.
GeometryUtils::ShapeFunctionsGradients needs the **local DN_De** (and InvJ0) to be available, and there is no other way to get DN_De at the integration point rather than calling **r_geometry.ShapeFunctionsLocalGradients**.

Reply:
But why don't you call the `GeometryUtils::CalculateGeometryData` which calculates both the shape functions and the gradients? This would be much more efficient as these are hardcoded for triangle and tetrahedra elements.

Reply:
Isn't it for a single integration (Gauss) point?

Reply:
For the midpoint. But as long as we only use linear elements the gradients are constant.

Reply:
You are right about the gradients. Anyway, I prefer to keep it this (more general) way to be consistent with nodal gradient and divergence processes and maybe different types of elements. But, if you think it is necessary to change it, let me know to use GeometryUtils::CalculateGeometryData for gradients (keeping the current code for shape functions).

Reply:
The point is this is highly inefficient. I already changed it while doing the TLS stuff. Different types of elements cannot be used as this would require implementing all the level set technology for everything.

Reply:
With last commit we are using the TLS. On top of that I added the more efficient gradients calculation. This required using stack arrays, which are more efficient too.

Reply:
Thank you @rubenzorrilla . Just let me know if I should do any modifications before merging.

Reply:
I don't think so. But I'd appreciate if you can try one of your cases (just to be completely sure that I didn't mess something).

Reply:
I checked a simple case. It seems to be Ok.